controller/src/main/java/io/pravega/controller/server/SegmentStoreConnectionManager.java

```diff
@@ -117,21 +118,66 @@ public void close() throws IOException {
             }, connectionObj -> connectionObj.connection.close(), maxConcurrent, maxIdle);
         }
 
-        CompletableFuture<ConnectionObject> getConnection(ReplyProcessor replyProcessor) {
+        CompletableFuture<ConnectionWrapper> getConnection(ReplyProcessor replyProcessor) {
             return getResource()
-                    .thenApply(connectionObject -> {
+                    .thenApply(closableResource -> {
+                        ConnectionObject connectionObject = closableResource.getResource();
                         connectionObject.reusableReplyProcessor.initialize(replyProcessor);
-                        return connectionObject;
+                        return new ConnectionWrapper(closableResource);
                     });
         }
-        
-        void returnConnection(ConnectionObject connectionObject) {
-            connectionObject.reusableReplyProcessor.uninitialize();
-            returnResource(connectionObject, connectionObject.state.get().equals(ConnectionObject.ConnectionState.CONNECTED));
+    }
+    
+    static class ConnectionWrapper implements AutoCloseable {
+        private final ResourcePool.ClosableResource<ConnectionObject> resource;
+        @GuardedBy("$lock")
+        private boolean isClosed;
+        private ConnectionWrapper(ResourcePool.ClosableResource<ConnectionObject> resource) {
+            this.resource = resource;
+            this.isClosed = false;
+        }
+
+        void failConnection() {
+            resource.getResource().failConnection();
+        }
+
+        <T> void sendAsync(WireCommand request, CompletableFuture<T> resultFuture) {
+            resource.getResource().sendAsync(request, resultFuture);
+        }
+
+        // region for testing
+        @VisibleForTesting
+        ConnectionObject.ConnectionState getState() {
+            return resource.getResource().state.get();
+        }
+
+        @VisibleForTesting
+        ClientConnection getConnection() {
+            return resource.getResource().connection;
+        }
+
+        @VisibleForTesting
+        ReplyProcessor getReplyProcessor() {
+            return resource.getResource().reusableReplyProcessor.replyProcessor.get();
+        }
+        // endregion
+
+        @Override
+        @Synchronized
+        public void close() {
```

Comment:
You're also holding a lock here while closing your resource, which in turn may complete a future while holding it.

Reply:
fixed